### PR TITLE
Add Abuse reporting

### LINKED ISSUE #5595: API endpoint to report an add-on/flag an add-on
The AMO frontend will soon have a "report a problem" button (we didn't have one before) but there is no API for reporting an issue with an add-on so we'll need one to implement that feature.

### PR DESCRIPTION
**Pull request by eviljeff**
fixes #5595

### CODE REVIEW

**Comment by eviljeff** [comment]
As this API doesn't require auth and creates stuff I'm not sure if we need throttling or otherwise to stop abuse (heh)

File: src/olympia/abuse/tests/test_views.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+    def setUp(self):
+        self.url = reverse('abusereport-list')
+
+    def check_reporter(self, report):
+        raise NotImplementedError
+
+    def check_report(self, report, text):
+        assert unicode(report) == text
+        assert report.ip_address
```
Comment (diox) [inline src/olympia/abuse/tests/test_views.py]: Can you actually check the value? You should be able to pass a custom `REMOTE_ADDR` to your `client.post()` calls.

File: docs/topics/api/abuse.rst [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    .. _abusereport-create-request:
+
+    :<json string|null addon: if reporting an add-on then the add-on id, slug, or guid of the add-on.
+    :<json string|null user: if reporting a user then the user id or username of the user.
+    :<json string message: An explanation of the reason for the report (required).
+    :>json object|null reporter: The user who submitted the report, if authenticated.
+    :>json int reporter.id: The id of the user who submitted the report.
+    :>json string reporter.name: The name of the user who submitted the report.
+    :>json string reporter.url: The link to the profile page for of the user who submitted the report.
+    :>json string ip_address: The IP address where the report was submitted from, to track potential abuse of this API.
```
Comment (diox) [inline docs/topics/api/abuse.rst]: I'm not sure this should be part of the response. It should certainly be available to staff later on when/if we expand the API and add a detail/list view, but I don't think it should be part of the response after a post.

Comment (tofumatt) [inline docs/topics/api/abuse.rst]: Agreed. I don't think it needs to be available by default.

File: docs/topics/api/abuse.rst [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+The following API endpoint covers abuse reporting
+
+-------------------
+Submitting a report
+-------------------
+
+.. _`abusereport-create`:
+
+The following API endpoint allows an abuse report to be submitted for an Add-on
+or user on addons.mozilla.org.  Authentication is not required, but is recommended
```
Comment (tofumatt) [inline docs/topics/api/abuse.rst]: Pinging @jvillalobos because requiring authentication to report abuse would likely:

1. lessen spam
2. encourage user sign-in/account creation
3. improve quality of abuse reports/follow-up

the only downside being it's harder to report. But we make reviews (arguably more important) require sign-in. Why not do it here? I know spam is a HUGE issue for this right now and I feel this would help a lot AND improve quality.

File: docs/topics/api/abuse.rst [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+so reports can be responded to if nessecary.
+
+.. http:post:: /api/v3/abuse/report/
+
+    .. _abusereport-create-request:
+
+    :<json string|null addon: if reporting an add-on then the add-on id, slug, or guid of the add-on.
```
Comment (tofumatt) [inline docs/topics/api/abuse.rst]: I would rephrase this to "add-on id, slug, or guid if the abuse report relates to an add-on".

But which one is it? It's confusing that it could be any.

File: docs/topics/api/abuse.rst [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    :<json string|null user: if reporting a user then the user id or username of the user.
```
Comment (tofumatt) [inline docs/topics/api/abuse.rst]: Same as above; switch it around so the if comes after.

File: docs/topics/api/abuse.rst [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    :<json string message: An explanation of the reason for the report (required).
```
Comment (tofumatt) [inline docs/topics/api/abuse.rst]: I think it can just be "the body/content of the abuse report".

File: docs/topics/api/abuse.rst [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    :>json object|null reporter: The user who submitted the report, if authenticated.
+    :>json int reporter.id: The id of the user who submitted the report.
+    :>json string reporter.name: The name of the user who submitted the report.
+    :>json string reporter.url: The link to the profile page for of the user who submitted the report.
```
Comment (tofumatt) [inline docs/topics/api/abuse.rst]: Will this work if they aren't an author? The plan is only to show profiles for users who submitted reviews/add-ons. So a lot of these would 404.

File: docs/topics/api/abuse.rst [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    :>json string ip_address: The IP address where the report was submitted from, to track potential abuse of this API.
+    :>json object|null addon: If this was an add-on abuse report, then the add-on.
+    :>json string addon.guid: The add-on `extension identifier <https://developer.mozilla.org/en-US/Add-ons/Install_Manifests#id>`_.
+    :>json int|null addon.id: The add-on id on AMO.  If the guid submitted didn't match a known add-on on AMO, then null.
```
Comment (tofumatt) [inline docs/topics/api/abuse.rst]: Did you type this on a typewriter? Why the double space?! ⌨️ 😅

File: docs/topics/api/abuse.rst [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    :>json string|null addon.slug: The add-on slug.  If the guid submitted didn't match a known add-on on AMO, then null.
+    :>json object|null user: If this was a user abuse report, then the user.
```
Comment (tofumatt) [inline docs/topics/api/abuse.rst]: "User reported for abuse" maybe is better phrasing (ironically I phrased this comment terribly)?

File: docs/topics/api/abuse.rst (hunk 3 above)
Comment (jvillalobos) [inline docs/topics/api/abuse.rst]: While reducing the noise in abuse reports would be good, I don't think requiring authentication is the way to go there. We get useful info from anonymous reports, and encouraging account creation is not a goal for AMO.

We do have some ideas to improve on this in the near future, in particular direct support requests, ratings, and abuse reports to where they belong, and have some predetermined abuse report options instead of just an open text field.

File: docs/topics/api/abuse.rst (hunk 4 above)
Comment (eviljeff) [inline docs/topics/api/abuse.rst]: It's like the add-on apis, they accept all three (this viewset uses AddonViewSet to get the Addon from input).

File: docs/topics/api/abuse.rst (hunk 7 above)
Comment (eviljeff) [inline docs/topics/api/abuse.rst]: We'd not started on #5592 so I just used the same serializer we use elsewhere (which does include the profile link).  I guess it makes sense to stop including it in new APIs.

File: docs/topics/api/abuse.rst (hunk 8 above)
Comment (eviljeff) [inline docs/topics/api/abuse.rst]: They'll take my double space from my cold dead hands!

**Comment by eviljeff** [comment]
> Also, what if we had separate endpoints for addon and user abuse reports? I know historically they're the same model but the API could abstract that, it's kind of weird they're linked and that we'll often have responses with half-null data sets.

So it'd be:
/api/v3/abuse/reportaddon/
/api/v3/abuse/reportuser/
or 
/api/v3/abuse/report/addon/ 
/api/v3/abuse/report/user/

You still want the addon ID in the POST, not the url?

File: docs/topics/api/abuse.rst (hunk 7 above)
Comment (eviljeff) [inline docs/topics/api/abuse.rst]: Actually, thinking further on this - it's going to more straightforward to keep it as is for now and then make the change everywhere at once.  The frontend should be able to trust that if a profile url has been returned it will be accessible.

File: src/olympia/abuse/tests/test_views.py (hunk 1 above)
Comment (eviljeff) [inline src/olympia/abuse/tests/test_views.py]: https://github.com/mozilla/addons-server/pull/5975/commits/2f78fb22c3f2898d91a2c29c2799e3142b073cdc

File: docs/topics/api/abuse.rst (hunk 2 above)
Comment (eviljeff) [inline docs/topics/api/abuse.rst]: https://github.com/mozilla/addons-server/pull/5975/commits/2f78fb22c3f2898d91a2c29c2799e3142b073cdc

**Comment by eviljeff** [comment]
https://github.com/mozilla/addons-server/pull/5975/commits/5f22b8583d5269719e4664991aa059786ae9a847 splits the endpoints, and changes the docs as appropriate (some of the requested changes were no longer required, because there were a whole lot less ifs going on)

File: docs/topics/api/abuse.rst [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+---------------------------------
+
+.. _`addonabusereport-create`:
+
+The following API endpoint allows an abuse report to be submitted for an Add-on,
+either listed on https://addons.mozilla.org or not.
+Authentication is not required, but is recommended so reports can be responded
+to if nessecary.
+
+.. http:post:: /api/v3/abuse/reportaddon/
```
Comment (tofumatt) [inline docs/topics/api/abuse.rst]: `/api/v3/abuse/report-addon/` or `/api/v3/abuse/report/addon/` would be easier to read/cleaner... and I think using the latter makes more sense as a report API then has subresources (`addon` and `user`) and the API should separate them out.

File: docs/topics/api/abuse.rst [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+    .. _addonabusereport-create-request:
+
+    :<json string addon: The add-on id, slug, or guid of the add-on (required).
```
Comment (tofumatt) [inline docs/topics/api/abuse.rst]: "add-on" is in this sentence twice. Maybe just "The id, slug, or guid of the add-on to report for abuse."

File: docs/topics/api/abuse.rst [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    :<json string message: The body/content of the abuse report (required).
+    :>json object|null reporter: The user who submitted the report, if authenticated.
+    :>json int reporter.id: The id of the user who submitted the report.
+    :>json string reporter.name: The name of the user who submitted the report.
```
Comment (tofumatt) [inline docs/topics/api/abuse.rst]: Is this stuff checked against the AMO users DB or is it just submitted freeform and we're expected to use the right data? If it's checked it seems we should really only need to submit a user id, username, or email and AMO can fetch the rest of this. Seems silly to submit it all.

File: docs/topics/api/abuse.rst [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+    :>json string reporter.url: The link to the profile page for of the user who submitted the report.
+    :>json object addon: The add-on reported for abuse.
```
Comment (tofumatt) [inline docs/topics/api/abuse.rst]: Above it says "addon" is a string (id/guid/slug) but here it's an object. Which is it? 😄

File: docs/topics/api/abuse.rst [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+    :>json string addon.guid: The add-on `extension identifier <https://developer.mozilla.org/en-US/Add-ons/Install_Manifests#id>`_.
+    :>json int|null addon.id: The add-on id on AMO.  If the guid submitted didn't match a known add-on on AMO, then null.
```
Comment (tofumatt) [inline docs/topics/api/abuse.rst]: Double space here, it's just not how we tend to write docs.

File: docs/topics/api/abuse.rst [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+    :>json string|null addon.slug: The add-on slug.  If the guid submitted didn't match a known add-on on AMO, then null.
```
Comment (tofumatt) [inline docs/topics/api/abuse.rst]: Double space here, it's just not how we tend to write docs.

File: docs/topics/api/abuse.rst [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+    :>json string message: The body/content of the abuse report.
```
Comment (tofumatt) [inline docs/topics/api/abuse.rst]: This seems to be here twice.

File: docs/topics/api/abuse.rst [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+Submitting a user abuse report
+------------------------------
+
+.. _`userabusereport-create`:
+
+The following API endpoint allows an abuse report to be submitted for a user account
+on https://addons.mozilla.org.  Authentication is not required, but is recommended
+so reports can be responded to if nessecary.
+
+.. http:post:: /api/v3/abuse/reportuser/
```
Comment (tofumatt) [inline docs/topics/api/abuse.rst]: Similarly, `/api/v3/abuse/report/user/` I think would be better.

File: docs/topics/api/abuse.rst [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+
+    .. _userabusereport-create-request:
+
+    :<json string user: The user id or username of the user (required).
```
Comment (tofumatt) [inline docs/topics/api/abuse.rst]: user id seems redundant, "id" should be fine.

File: docs/topics/api/abuse.rst [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+    :<json string message: The body/content of the abuse report (required).
+    :>json object|null reporter: The user who submitted the report, if authenticated.
+    :>json int reporter.id: The id of the user who submitted the report.
+    :>json string reporter.name: The name of the user who submitted the report.
+    :>json string reporter.url: The link to the profile page for of the user who submitted the report.
+    :>json object user: The user reported for abuse.
+    :>json int user.id: The id of the user reported.
```
Comment (tofumatt) [inline docs/topics/api/abuse.rst]: Again, seems weird to include all this info when it's on AMO. Couldn't the user ID be enough?

File: docs/topics/api/abuse.rst [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+    :>json string user.name: The name of the user reported.
+    :>json string user.url: The link to the profile page for of the user reported.
+    :>json string message: The body/content of the abuse report.
```
Comment (tofumatt) [inline docs/topics/api/abuse.rst]: This seems duplicated as well.

**Comment by eviljeff** [comment]
As discussed on irc, I believe the sphinx syntax in the docs has been misinterpreted so I edited all the comments that appear to conflate request and response content.

File: docs/topics/api/abuse.rst (hunk 10 above)
Comment (eviljeff) [inline docs/topics/api/abuse.rst]: https://github.com/mozilla/addons-server/pull/5975/commits/7596c05a0b4550bb5f8b90b854daa1aaf01a8bf7

File: docs/topics/api/abuse.rst (hunk 17 above)
Comment (eviljeff) [inline docs/topics/api/abuse.rst]: https://github.com/mozilla/addons-server/pull/5975/commits/7596c05a0b4550bb5f8b90b854daa1aaf01a8bf7

**Comment by tofumatt** [comment]
Thanks, sorry I misread the docs yesterday. fwiw r+ from me too 😄
